libs/s25main/desktops/dskCampaignMissionSelection.cpp

```diff
@@ -0,0 +1,197 @@
+dskCampaignMissionSelection::dskCampaignMissionSelection(CreateServerInfo csi, boost::filesystem::path campaignFolder)
+    : Desktop(LOADER.GetImageN("setup015", 0)), campaignFolder_(std::move(campaignFolder)), csi_(std::move(csi)),
+      currentPage_(0), lastPage_(0), missionsPerPage_(10)
+{
+    const unsigned int btOffset = getStartOffsetMissionButtonsY()
+                                  + missionsPerPage_ * (missionButtonHeightY + distanceBetweenMissionButtonsY)
+                                  + distanceBetweenElementsY;
+    AddTextButton(ID_Back, DrawPoint(300, 560), Extent(200, 22), TextureColor::Red1, _("Back"), NormalFont);
+
+    auto* imageFirstPage = LOADER.GetImageN("io", 102);
+    AddImageButton(ID_FirstPage, DrawPoint(400 - buttonWidth * 3 - 2 * spacingBetweenButtons, btOffset),
+                   Extent(buttonWidth, buttonHeight), TextureColor::Green2, imageFirstPage);
+
+    auto* imagePreviousPage = LOADER.GetImageN("io", 103);
+    AddImageButton(ID_PreviousPage, DrawPoint(400 - buttonWidth * 2, btOffset), Extent(buttonWidth, buttonHeight),
+                   TextureColor::Green2, imagePreviousPage);
+
+    auto* imageNextPage = LOADER.GetImageN("io", 104);
+    AddImageButton(ID_NextPage, DrawPoint(400 + buttonWidth, btOffset), Extent(buttonWidth, buttonHeight),
+                   TextureColor::Green2, imageNextPage);
+
+    auto* imagelastPage_ = LOADER.GetImageN("io", 105);
+    AddImageButton(ID_LastPage, DrawPoint(400 + buttonWidth * 2 + 2 * spacingBetweenButtons, btOffset),
+                   Extent(buttonWidth, buttonHeight), TextureColor::Green2, imagelastPage_);
+
+    settings_ = std::make_unique<CampaignDescription>();
+    CampaignDataLoader loader(*settings_, campaignFolder_);
+    if(!loader.Load() || settings_->mapNames.empty())
+        LOG.write(_("Failed to load campaign %1%.\n")) % campaignFolder_;
+
+    lastPage_ = (settings_->mapNames.size() - 1) / missionsPerPage_;
+
+    UpdateEnabledStateOfNextPreviousButton();
+
+    AddText(ID_PageLabel, DrawPoint(400, btOffset + buttonHeight / 2),
+            std::to_string(currentPage_ + 1) + "/" + std::to_string(lastPage_ + 1), COLOR_YELLOW,
+            FontStyle::CENTER | FontStyle::VCENTER, LargeFont);
+
+    AddText(ID_ChooseCapitalLabel, DrawPoint(400, startOffsetY), _("Choose capital"), COLOR_YELLOW, FontStyle::CENTER,
+            LargeFont);
+
+    UpdateMissionPage();
+}
+
+void dskCampaignMissionSelection::UpdateMissionPage()
+{
+    ctrlGroup* group = AddGroup(ID_GroupStart + currentPage_);
+    Extent catBtSize = Extent(300, missionButtonHeightY);
+    DrawPoint curBtPos(250, getStartOffsetMissionButtonsY());
+    for(unsigned int i = 0; i < missionsPerPage_; i++)
+    {
+        unsigned int missionIndex = currentPage_ * missionsPerPage_ + i;
+        if(missionIndex >= settings_->mapNames.size())
+            break;
+
+        const bfs::path& mapFilePath = RTTRCONFIG.ExpandPath(settings_->mapFolder) / settings_->mapNames[missionIndex];
```

Comment:
At least use the function (and get rid of the RTTRConfig include):
```suggestion
        const bfs::path& mapFilePath = settings_->getMapFilePath(settings_->mapNames[missionIndex]);
```

That's also why I thought about making `mapNames/luaFolder/mapFolder` private and provide only access to the map count and map/lua file path by index. Iterating over it (instead of mapNames) could be done by `for(auto idx: helpers::range(settings.getNumMaps()))` but here you already have indices

Reply:
Potential i will implement this :)

Reply:
Added index access and made mapNames/luaFolder/mapFolder private. And use it at calling side.